### DISCUSSION TITLE
With multiple team members, what is an efficient way to collaborate with mkdocs-material-insiders?

### DISCUSSION THREAD

**Discussion by ejkim-morai** [squidfunk/mkdocs-material#5756] [Q&A] [answered]
Hello. I sponsored Indsider with my personal account and had access to the mkdocs-material-insiders Private Repo(https://github.com/squidfunk/mkdocs-material-insiders).
Now I want to collaborate with my team by sponsoring Indsider with the organizational account I created.
My team members need to do `pip install git+https://${GH_TOKEN}@github.com/squidfunk/mkdocs-material-insiders.git` as specified in your guide(https://squidfunk.github.io/mkdocs-material/insiders/getting-started/) to locally install and build the repo with Insider installed, but they don't have access to your repo, so they can't seem to install it. (Even though each person has a personal access token)
I think it's inefficient to ask you on a case-by-case basis for all team members on my organizational account to have access to the Insider Repo.

In inquiry #3816(https://github.com/squidfunk/mkdocs-material/discussions/3816) in Discussions, I found that you said "The general advice for organizations is to fork the Insiders repository to your organization, which allows you to grant access to all members of your organization".
I would like a detailed guide on this.
Actually, I forked the mkdocs-material-insiders  Private Repo to my organization account, but my teammates seem to be unable to access the forked repo in my organization account.

**Comment by squidfunk** [comment, marked as answer]
Forking is the way to go, and it's what all our organizational sponsors do. Note that you might need to adjust the rights of your teammates to access the forked repository, as it's a private repository in your organization. [See GitHub docs here](https://docs.github.com/en/organizations/managing-user-access-to-your-organizations-repositories/managing-an-individuals-access-to-an-organization-repository). You teammates can then create a personal access token (PAT) to access the repository.

**Comment by ejkim-morai** [comment]
Thank you for helping me.
As you guided, I added my colleague to the forked private repo and was able to install Indsider with the PAT my colleague received.

On Windows, I was able to build the document by installing Insider in the following order.
1. pip install mkdocs-material
2. python.exe -m pip install git+https://{ghp_token}@github.com/{git repo name}/mkdocs-material-insiders.git
3. Install additional plugins

However, after installing the same in WSL, 
the following !ENV-related error occurred when building the document with `mkdocs serve`.
```
NFO    -  Building documentation...
Traceback (most recent call last):
  File "/home/ejkim/.local/bin/mkdocs", line 11, in <module>
    sys.exit(cli())
  File "/usr/lib/python3/dist-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3/dist-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python3/dist-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3/dist-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/mkdocs/__main__.py", line 121, in serve_command
    serve.serve(
  File "/usr/lib/python3/dist-packages/mkdocs/commands/serve.py", line 110, in serve
    config = builder()
  File "/usr/lib/python3/dist-packages/mkdocs/commands/serve.py", line 96, in builder
    config = load_config(
  File "/usr/lib/python3/dist-packages/mkdocs/config/base.py", line 177, in load_config
    cfg.load_file(config_file)
  File "/usr/lib/python3/dist-packages/mkdocs/config/base.py", line 123, in load_file
    return self.load_dict(utils.yaml_load(config_file))
  File "/usr/lib/python3/dist-packages/mkdocs/utils/__init__.py", line 71, in yaml_load
    return yaml.load(source, Loader)
  File "/usr/lib/python3/dist-packages/yaml/__init__.py", line 72, in load
    return loader.get_single_data()
  File "/usr/lib/python3/dist-packages/yaml/constructor.py", line 37, in get_single_data
    return self.construct_document(node)
  File "/usr/lib/python3/dist-packages/yaml/constructor.py", line 46, in construct_document
    for dummy in generator:
  File "/usr/lib/python3/dist-packages/yaml/constructor.py", line 398, in construct_yaml_map
    value = self.construct_mapping(node)
  File "/usr/lib/python3/dist-packages/yaml/constructor.py", line 204, in construct_mapping
    return super().construct_mapping(node, deep=deep)
  File "/usr/lib/python3/dist-packages/yaml/constructor.py", line 129, in construct_mapping
    value = self.construct_object(value_node, deep=deep)
  File "/usr/lib/python3/dist-packages/yaml/constructor.py", line 86, in construct_object
    data = constructor(self, node)
  File "/usr/lib/python3/dist-packages/yaml/constructor.py", line 412, in construct_undefined
    raise ConstructorError(None, None,
yaml.constructor.ConstructorError: could not determine a constructor for the tag '!ENV'
  in "/home/ejkim/mkdocs-material/mkdocs.yml", line 101, column 15
```
  Looking at the error message, it seems that the module in the python3 path is being loaded.
 When I checked the pip version mapped to python3 with `python3 -m pip --version`, I saw the message below.
`/usr/bin/python3: No module named pip`
 This was the same when I checked with `python -m pip --version`.
(The python3 command seems to run Python 3.6.9, and the python command runs Python 3.8.)

Therefore, I was able to build the documentation by running python 3.11 with the `python.exe` command as shown below.
`python.exe -m mkdocs serve`
And I checked **pip 23.2.1** version with `python.exe -m pip --version`.

What I'm curious about
When I checked at https://pypi.org/project/mkdocs-material/, the requirement was that the python version should be 3.7 or higher, so I installed Python 3.8  and I made python 3.8 run by default with `Update-alternatives`
 But Why can't I find an executable pip like below?
  `/usr/bin/python: No module named pip'
And is the !ENV error related to the Python version?

thank you!

**Reply by squidfunk** [reply]
Not sure, as I'm not a Windows user. It might be a good idea to create a new discussion for this very problem, so maybe other users can help you resolve this 😊

**Reply by kamilkrzyskow** [reply]
WSL is the Linux subsystem, I guess you're using Ubuntu, as that's the default WSL distro.
Haven't used Linux in quite a while, but this website https://careerkarma.com/blog/python-pip-command-not-found/ claims that:

"On Linux, the pip package manager is an independent package. This means you must install pip separately from Python. On Mac, you do not need to worry about installing pip manually, as long as you are working with Python 3.x."

And I have, a vague memory that I indeed had to install the pip manager separately 🤷 

As for the error:
```
yaml.constructor.ConstructorError: could not determine a constructor for the tag '!ENV'
  in "/home/ejkim/mkdocs-material/mkdocs.yml", line 101, column 15
```

!ENV support was added in MkDocs 1.2, maybe you had an older version installed, hence the error.
Also I recommend always using virtual environments to manage your python packages across projects. I use the default built-in `python -m venv venv` :v:

**Comment by ejkim-morai** [comment]
Thank you for your valuable advice. In particular, this phrase **'On Linux, the pip package manager is an independent package. This means you must install pip separately from Python'** was a great help!
And when I checked as below, it seems that both python 3.6 and 3.8 versions do not include pip.
```
ejkim@MORAI-PC-075:~/morai-techblog.github.io$ python3 -m ensurepip --upgrade
/usr/bin/python3: No module named ensurepip
ejkim@MORAI-PC-075:~/morai-techblog.github.io$ python -m ensurepip --upgrade
/usr/bin/python: No module named ensurepip
```
In this case, I have to install the ensurepip module separately, which seems quite troublesome.
I think that the lower python version can't load pip and gives an error related to `!ENV`.

In conclusion, it would be better to use the latest version of python (3.11), which includes the pip module, or use venv like your recommendation.
My colleague also said that using 'venv' is the best.
thank you :)

**Reply by squidfunk** [reply]
My god, sometimes Python is so unnecessarily complicated 😅 Glad you fixed it!
